dex/ws/wslink.go

```diff
@@ -0,0 +1,262 @@
+		select {
+		case <-c.quit:
+			break out
+		default:
+		}
+		// Block until a message is received or an error occurs.
+		_, msgBytes, err := c.conn.ReadMessage()
+		if err != nil {
+			// Log the error if it's not due to disconnecting.
+			if err != io.EOF {
+				log.Errorf("Websocket receive error from %s: %v", c.ip, err)
+			}
+			break out
+		}
+		// Attempt to unmarshal the request. Only requests that successfully decode
+		// will be accepted by the server, though failure to decode does not force
+		// a disconnect.
+		msg := new(msgjson.Message)
+		err = json.Unmarshal(msgBytes, msg)
+		if err != nil {
+			c.SendError(1, msgjson.NewError(msgjson.RPCParseError,
+				"Failed to parse message: "+err.Error()))
+			continue
+		}
+		if msg.ID == 0 {
+			c.SendError(1, msgjson.NewError(msgjson.RPCParseError, "request id cannot be zero"))
+			continue
+		}
+		rpcErr := c.handler(msg)
+		if rpcErr != nil {
+			c.SendError(msg.ID, rpcErr)
+		}
+	}
+	// Ensure the connection is closed.
+	c.Disconnect()
+	c.wg.Done()
+}
+
+// outHandler handles all outgoing messages for the websocket connection.
+// It uses a buffered channel to serialize output messages while allowing the
+// sender to continue running asynchronously.  It must be run as a goroutine.
+func (c *WSLink) outHandler() {
+	ticker := time.NewTicker(c.pingPeriod)
+	ping := []byte{}
+out:
+	for {
+		// Send any messages ready for send until the quit channel is
+		// closed.
+		select {
+		case b := <-c.outChan:
+			err := c.conn.WriteMessage(websocket.TextMessage, b)
+			if err != nil {
+				c.Disconnect()
+				break out
+			}
+		case <-ticker.C:
+			err := c.conn.WriteMessage(websocket.PingMessage, ping)
+			if err != nil {
+				c.Disconnect()
+				// Don't really care what the error is, but log it at debug level.
+				log.Debugf("WriteMessage ping error: %v", err)
+				break out
+			}
+		case <-c.quit:
+			break out
+		}
+	}
+
+	// Drain any wait channels before exiting so nothing is left waiting
+	// around to send.
+cleanup:
+	for {
+		select {
+		case <-c.outChan:
+		default:
+			break cleanup
+		}
+	}
+	c.wg.Done()
+	log.Tracef("Websocket client output handler done for %s", c.ip)
+}
+
+// Off will return true if the client has disconnected.
+func (c *WSLink) Off() bool {
+	c.quitMtx.RLock()
+	defer c.quitMtx.RUnlock()
+	return !c.on
+}
+
+// IP is the address passed to the constructor.
+func (c *WSLink) IP() string {
+	return c.ip
+}
+
+func NewConnection(w http.ResponseWriter, r *http.Request, wait time.Duration) (Connection, error) {
+	ws, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		if _, ok := err.(websocket.HandshakeError); !ok {
+			return nil, fmt.Errorf("Unexpected websocket error: %v", err)
+		}
+	}
+	pongHandler := func(string) error {
+		return ws.SetReadDeadline(time.Now().Add(wait))
+	}
+	err = pongHandler("")
+	if err != nil {
+		return nil, fmt.Errorf("error setting read deadline: %v", err)
+	}
+	ws.SetPongHandler(pongHandler)
+	return ws, nil
+}
```

Comment:
<s>Any possibility to link this with the previous file via `git mv` or would it turn out the same when squashed?</s>  Nvm there's a fair amount left in server/comms/link.go.